vespa-hadoop/src/main/java/com/yahoo/vespa/hadoop/mapreduce/VespaRecordWriter.java

```diff
@@ -0,0 +1,146 @@
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.logging.Logger;
+
+import static java.util.stream.Collectors.toList;
+
+/**
+ * {@link VespaRecordWriter} sends the output &lt;key, value&gt; to one or more Vespa endpoints using vespa-feed-client.
+ *
+ * @author bjorncs
+ */
+public class VespaRecordWriter extends RecordWriter<Object, Object> {
+
+    private final static Logger log = Logger.getLogger(VespaRecordWriter.class.getCanonicalName());
+
+    private final VespaCounters counters;
+    private final VespaConfiguration config;
+
+    private boolean initialized = false;
+    private JsonStreamFeeder feeder;
+
+    protected VespaRecordWriter(VespaConfiguration config, VespaCounters counters) {
+        this.counters = counters;
+        this.config = config;
+    }
+
+    @Override
+    public void write(Object key, Object data) throws IOException {
+        initializeOnFirstWrite();
+        String json = "[" + data.toString().trim() + "]";
+        // TODO Feed asynchronous
+        try {
+            counters.incrementDocumentsSent(1);
+            feeder.feed(new ByteArrayInputStream(json.getBytes(StandardCharsets.UTF_8)));
+            counters.incrementDocumentsOk(1);
+        } catch (Exception e) {
+            // TODO Document id and more details in error message
+            log.warning("Failed to feed single document: " + e);
+            counters.incrementDocumentsFailed(1);
+        }
+        if (counters.getDocumentsSent() % config.progressInterval() == 0) {
+            String progress = String.format("Feed progress: %d / %d / %d / %d (sent, ok, failed, skipped)",
+                    counters.getDocumentsSent(),
+                    counters.getDocumentsOk(),
+                    counters.getDocumentsFailed(),
+                    counters.getDocumentsSkipped());
+            log.info(progress);
+        }
+    }
+
+    @Override
+    public void close(TaskAttemptContext context) throws IOException {
+        if (feeder != null) {
+            feeder.close();
+            feeder = null;
+            initialized = false;
+        }
+    }
+
+    /** Override method to alter {@link FeedClient} configuration */
+    protected void onFeedClientInitialization(FeedClientBuilder builder) {}
+
+    private void initializeOnFirstWrite() {
+        if (initialized) return;
+        validateConfig();
+        useRandomizedStartupDelayIfEnabled();
+        feeder = createJsonStreamFeeder();
+        initialized = true;
+    }
+
+    private void validateConfig() {
+        if (!config.useSSL()) {
+            throw new IllegalArgumentException("SSL is required for this feed client implementation");
+        }
+        if (config.dryrun()) {
+            throw new IllegalArgumentException("Dryrun is not supported for this feed client implementation");
```

Comment:
We should probably add a dry-run option.

Reply:
I'll add this once we are settled on a http client implementation

Reply:
Added in b525d31